### PR TITLE
New parameter `mirror_type`

### LINKED ISSUE #632: Add new parameter: mirror_type: custom, none
Right now, when doing a module_type: custom, we still get the mirror goal, which is often just wrong, for example when the custom module does not need a mirror. (ontofox, wikidata API, etc)

### PR DESCRIPTION
**Pull request by anitacaron**
Fixes #632

### CODE REVIEW

File: template/src/ontology/Makefile.jinja2 [hunk 1]
```diff
@@ -444,7 +444,7 @@ $(IMPORTDIR)/{{ ont.id }}_import.owl: {% if project.import_group.use_base_mergin
 		remove $(patsubst %, --term %, $(ANNOTATION_PROPERTIES)) -T $(IMPORTDIR)/{{ ont.id }}_terms_combined.txt --select complement --select "classes individuals annotation-properties" \
 		$(ANNOTATE_CONVERT_FILE); fi
 {% elif ('custom' == ont.module_type) or (ont.module_type is none and 'custom' == project.import_group.module_type) %}
-$(IMPORTDIR)/{{ ont.id }}_import.owl: {% if project.import_group.use_base_merging %}$(MIRRORDIR)/merged.owl{% else %}$(MIRRORDIR)/{{ ont.id }}.owl{% endif %}
+$(IMPORTDIR)/{{ ont.id }}_import.owl: {% if project.import_group.use_base_merging %}$(MIRRORDIR)/merged.owl{% elif 'none' != ont.mirror_type %}$(MIRRORDIR)/{{ ont.id }}.owl{% endif %}
```
Comment (matentzn) [inline template/src/ontology/Makefile.jinja2:447]: So what should happen here if mirror type is none? Does this mean there  _is no mirror_? If so its risky to make this default!

File: odk/odk.py [hunk 2]
```diff
@@ -150,6 +150,9 @@ class ImportProduct(Product):
     use_gzipped: bool = False
     """if use_gzipped is true, try use the base IRI instead of normal one to mirror from."""
 
+    mirror_type: Optional[str] = None
```
Comment (matentzn) [inline odk/odk.py:153]: is `None` here not something else then the string `'none'` you look for later?

File: template/src/ontology/Makefile.jinja2 [hunk 3]
```diff
@@ -535,13 +535,17 @@ IMP_LARGE=true # Global parameter to bypass handling of large imports
 
 {% for ont in project.import_group.products %}
 
+{% if ont.mirror_type != 'none' -%}
 ## ONTOLOGY: {{ ont.id }}
+{% endif -%}
```
Comment (matentzn) [inline template/src/ontology/Makefile.jinja2:540]: Is this "endif" at the right place? Why you you skop the ONTOLOGY comment but not the "description" comment?

File: template/src/ontology/Makefile.jinja2 [hunk 4]
```diff
@@ -566,6 +570,10 @@ mirror-{{ ont.id }}: | $(TMPDIR)
 		$(ROBOT) convert -i $(MIRRORDIR)/{{ ont.id }}.owl.gz -o $@.tmp.owl{%- if ont.make_base %} && \
 		$(ROBOT) remove -i $@.tmp.owl {% if ont.base_iris is not none %}{% for iri in ont.base_iris %}--base-iri {{iri}} {% endfor %}{% else %}--base-iri $(URIBASE)/{{ ont.id.upper() }} {% endif %}--axioms external --preserve-structure false --trim false -o $@.tmp.owl{% endif %} &&\
 		mv $@.tmp.owl $(TMPDIR)/$@.owl; fi
+{%- elif 'custom' == ont.module_type and 'custom' == ont.mirror_type %}
+mirror-{{ ont.id }}: 
+	echo "ERROR: You have configured your default mirror type to be custom; this behavior needs to be overwritten in {{ project.id }}.Makefile!" && false
+{%- elif 'custom' == ont.module_type and 'none' == ont.mirror_type -%}
```
Comment (matentzn) [inline template/src/ontology/Makefile.jinja2:576]: Why do you need to couple the `module_type` and the `mirror_type` here so tightly? Couldnt the be configured separately?

**Review by matentzn — CHANGES_REQUESTED** [review]
@anitacaron excellent thinking of designing a `mirror_type`; this is the most flexible and appropriate way to solve many of our problems, and I never though of it. Super! It will take a few rounds to get this right, hope you bear with me. Thanks!
 
- [x] The use of 'none' string is too scary in a makefile, because its not clear if this is a string of "undefined" (no variable set). Can we rename to `no_mirror`?
 - [x] The idea with mirror_type is great. I don't think you need to couple it with module_type though (by couple I mean, check them both). The mirror can be custom while the module_type is fixed, and the other way around. 
 - [ ] The natural continuation would be to implement `mirror_type: base` - don't remove `make_base: TRUE` but the same logic could apply here.
 - [ ] This PR should have no effect at all on (1) any currently implemented pipeline anywhere, and (2) on the imports goals.

File: odk/odk.py (hunk 2 above)
Comment (anitacaron) [reply odk/odk.py:153]: No. It's the meaning in the python. It's an optional string value, then the default is None. Same as `module_type`.

File: template/src/ontology/Makefile.jinja2 (hunk 1 above)
Comment (anitacaron) [reply template/src/ontology/Makefile.jinja2:447]: Yes, none means no mirror. Can you detail it more?

Comment (anitacaron) [reply template/src/ontology/Makefile.jinja2:447]: This is the main reason for the issue we want to fix, no?

File: template/src/ontology/Makefile.jinja2 (hunk 3 above)
Comment (anitacaron) [reply template/src/ontology/Makefile.jinja2:540]: I thought again, and it's better if the comments stay there with the comment below when there isn't a mirror.
I skipped only the phony/precious part.

File: template/src/ontology/Makefile.jinja2 (hunk 1 above)
Comment (anitacaron) [reply template/src/ontology/Makefile.jinja2:447]: If there isn't a mirror, it won't add the mirror file.

**Comment by anitacaron** [comment]
- I've changed the value `none` to `no_mirror`.

File: odk/odk.py (hunk 2 above)
Comment (matentzn) [reply odk/odk.py:153]: Ok

File: template/src/ontology/Makefile.jinja2 [hunk 5]
```diff
@@ -396,7 +396,7 @@ $(IMPORTDIR)/%_import.owl: {% if project.import_group.use_base_merging %}$(MIRRO
 		query --update ../sparql/inject-subset-declaration.ru --update ../sparql/inject-synonymtype-declaration.ru --update ../sparql/postprocess-module.ru \
 		$(ANNOTATE_CONVERT_FILE); fi
 {% elif 'custom' == project.import_group.module_type %}
-$(IMPORTDIR)/%_import.owl: {% if project.import_group.use_base_merging %}$(MIRRORDIR)/merged.owl{% else %}$(MIRRORDIR)/%.owl{% endif %}
+$(IMPORTDIR)/%_import.owl: {% if project.import_group.use_base_merging %}$(MIRRORDIR)/merged.owl{% elif 'custom' == ont.mirror_type %}$(MIRRORDIR)/%.owl{% endif %}
```
Comment (matentzn) [inline template/src/ontology/Makefile.jinja2:399]: I dont like this.. ODK should never change the default behaviour! Even if its `custom` module type, the mirror can be the default normal. I think the condition here should be 

```
{% elif 'no_mirror' != ont.mirror_type %}
```

Comment (matentzn) [reply template/src/ontology/Makefile.jinja2:399]: Correct me if I am wrong though!

**Review by matentzn — CHANGES_REQUESTED** [review]
- [x] Also add a row in CHANGELOG.md

File: template/src/ontology/Makefile.jinja2 (hunk 4 above)
Comment (anitacaron) [reply template/src/ontology/Makefile.jinja2]: I removed the check for both parameters.

File: template/src/ontology/Makefile.jinja2 (hunk 5 above)
Comment (anitacaron) [reply template/src/ontology/Makefile.jinja2:399]: You're right. This should be in sync with the other import part.

File: template/src/ontology/Makefile.jinja2 [hunk 6]
```diff
@@ -540,19 +540,21 @@ IMP_LARGE=true # Global parameter to bypass handling of large imports
...
 
+{%- if ont.mirror_type != 'no_mirror' -%}
 .PHONY: mirror-{{ ont.id }}
 .PRECIOUS: $(MIRRORDIR)/{{ ont.id }}.owl
+{%- endif -%}
 
 {%- if ont.mirror_from %}
 mirror-{{ ont.id }}: | $(TMPDIR)
-	if [ $(MIR) = true ] && [ $(IMP) = true ]{% if ont.is_large %} && [ $(IMP_LARGE) = true ]{% endif %}; then $(ROBOT) convert -I {{ ont.mirror_from }} -o $@.tmp.owl{%- if ont.make_base %} && \
+	if [ $(MIR) = true ] && [ $(IMP) = true ]{% if ont.is_large %} && [ $(IMP_LARGE) = true ]{% endif %}; then $(ROBOT) convert -I {{ ont.mirror_from }} -o $@.tmp.owl{%- if ont.make_base or 'base' == ont.mirror_type %} && \
```
Comment (matentzn) [inline template/src/ontology/Makefile.jinja2:550]: I am assuming this is the only time where `if ont.make_base` was needed..

Comment (matentzn) [reply template/src/ontology/Makefile.jinja2:550]: Oh foregt it.

**Review by matentzn — APPROVED** [review]

**Pull request merged**